src/9FC00B40.c

```diff
@@ -1,35 +1,113 @@
-#include "include_asm.h"
 #include "bbtypes.h"
+#include "libcrypto/aes.h"
+#include "libcrypto/sha1.h"
 #include "bcp.h"
 #include "macros.h"
 
+extern BbContentMetaDataHead contentMetaDataHead;
 extern BbTicketHead ticketHead;
 extern BbVirage2* virage2_offset;
-extern BbVirage01 D_9FC0F308;
-extern s16 D_9FC0EBB2;
-extern s32 D_9FC0EBB8;
+extern u16 D_9FC0EBB0;
+extern u16 D_9FC0EBB2;
+extern u32 D_9FC0EBB8;
 extern u32 D_9FC0ED30;
+extern u32 D_9FC0F0DC;
+extern AesKeyInstance D_9FC0F0E0;
+extern AesCipherInstance D_9FC0F2C8;
+extern u32 D_9FC0F2DC;
+extern u32 D_9FC0F2E0;
+extern BbAesIv D_9FC0F2E4;
 extern s32 D_9FC0F304;
+extern BbVirage01 D_9FC0F308;
 extern u16 g_cur_proc_trial_type;
-extern s32 g_trial_time_elapsed;
+extern u32 g_trial_time_elapsed;
+extern SHA1Context sha1_ctx;
 
 void startup(void);
-
+void launch_app_trampoline(void);
 void write_virage01_data(BbVirage01*);
-
-s32 check_cert_ranges(BbCertBase*);
-
-
+s32 check_cert_ranges(BbCertBase**);
+u16* getTrialConsumptionByCid(u16);
 
 s32 check_ticket_bundle_ranges(BbTicketBundle* bundle) {
     if (CHECK_UNTRUSTED(bundle) && CHECK_UNTRUSTED(bundle->ticket) && check_cert_ranges(bundle->ticketChain) &&
         check_cert_ranges(bundle->cmdChain)) {
-        return 1;
+        return TRUE;
     }
-    return 0;
+
+    return FALSE;
 }
 
-INCLUDE_ASM("asm/non_matchings/9FC00B40", func_9FC00BAC);
+s32 func_9FC00BAC(BbTicketBundle* bundle) {
+    BbAesKey key1;
+    rsaDataBlock blocks[2];
+    BbAesKey key2;
+    u16* cc;
+    s16 tid;
+    BbTicketHead* head = &bundle->ticket->head;
+    BbContentMetaDataHead* cmdHead = &bundle->ticket->cmd.head;
+
+    if (verify_cert_chain(bundle->cmdChain, 2) != 0) {
+        return -1;
+    }
+
+    if (verify_cert_chain(bundle->ticketChain, 1) != 0) {
+        return -1;
+    }
+
+    if (head->bbId != virage2_offset->bbId) {
+        return -1;
+    }
+
+    tid = head->tid;
+
+    if (tid < 0) {
+        cc = getTrialConsumptionByCid(head->tid);
+
+        if (cc == NULL) {
+            return -1;
+        }
+
+        if (head->code < 3 && *cc >= head->limit) {
+            return -1;
+        }
+    }
+
+    blocks[0].data = bundle->ticket;
+    blocks[0].size = sizeof(BbTicket) - sizeof(BbRsaSig2048);
+
+    if (rsa_verify_signature(
+        blocks,
+        1,
+        ((BbRsaCert*) bundle->ticketChain[0])->publicKey,
+        ((BbRsaCert*) bundle->ticketChain[0])->exponent,
+        0,
+        bundle->ticket->head.ticketSign) >= 0) {
```

Comment:
It's slightly difficult to tell at a glance where the condition ends and the block begins with this formatting, although I'm in agreement that it's too long to be on one line. I suggest indenting the other lines similarly to the following to make it a bit more obvious

```suggestion
    if (rsa_verify_signature(blocks, 1, ((BbRsaCert*) bundle->ticketChain[0])->publicKey,
                             ((BbRsaCert*) bundle->ticketChain[0])->exponent, 0,
                             bundle->ticket->head.ticketSign) >= 0) {
```

Reply:
was thinking about this. a possible alternative is using a local variable for the pointer casts eg

```c
cert = (BbRsaCert*) bundle->ticketChain[0];

if (rsa_verify_signature(blocks, 1, cert->publicKey, cert->exponent, 0, bundle->ticket->head.ticketSign) >= 0) {
```

Reply:
```c
    blocks[0].data = bundle->ticket;
    blocks[0].size = sizeof(BbTicket) - sizeof(BbRsaSig2048);

    if (rsa_verify_signature(
            blocks,
            1,
            ((BbRsaCert*) bundle->ticketChain[0])->publicKey,
            ((BbRsaCert*) bundle->ticketChain[0])->exponent,
            0,
            bundle->ticket->head.ticketSign) >= 0) {
        memcpy(&contentMetaDataHead, cmdHead, sizeof(BbContentMetaDataHead));
        eccGenAesKey(head->serverKey, virage2_offset->privateKey, key1);
        aes_SwDecrypt(key1, head->cmdIv, cmdHead->key, sizeof(BbAesKey), contentMetaDataHead.key);
```

dunno if better